torch/csrc/jit/codegen/cuda/kernel.cpp

```diff
@@ -358,6 +361,73 @@ std::vector<Expr*>& KernelInternalProxy::topLevelExprs() {
   return kernel_->top_level_exprs_;
 }
 
+void KernelPerformanceProfile::registerExpr(const Expr* expr) {
+  if (expr_entry_map_.find(expr) != expr_entry_map_.end()) {
+    return;
+  }
+
+  auto slot = getNewIndex();
+  expr_entry_map_.emplace(expr, slot);
+}
+
+int KernelPerformanceProfile::getNewIndex() {
+  auto slot = num_profile_entries_;
+  ++num_profile_entries_;
+  return slot;
+}
+
+bool KernelPerformanceProfile::isProfiled(const Expr* expr) const {
+  return expr_entry_map_.find(expr) != expr_entry_map_.end();
+}
+
+c10::optional<int> KernelPerformanceProfile::getIndex(const Expr* expr) const {
+  auto it = expr_entry_map_.find(expr);
+  if (it == expr_entry_map_.end()) {
+    return c10::optional<int>();
+  } else {
+    return it->second;
+  }
+}
+
+std::array<int, 2> KernelPerformanceProfile::getIndicesInProfileBuffer(
+    const Expr* expr) const {
+  if (!isProfiled(expr)) {
+    return {};
```

Comment:
Should you assert if it's not profiled? I assume std::array will initialize to `0, 0` here?

Reply:
Assertion seems more reasonable. Updated.